controller/src/main/java/io/pravega/controller/server/SegmentHelper.java

```diff
@@ -88,64 +103,77 @@ public SegmentHelper(final ConnectionFactory clientCF, HostControllerStore hostS
-                log.info(requestId, "CreateSegment {} segmentAlreadyExists", qualifiedStreamSegmentName);
-                result.complete(true);
+    private boolean isReplyWanted(Reply reply, Class[] classArray) {
+        for (Class<?> c : classArray) {
+            if (c.isAssignableFrom(reply.getClass())) {
+                return true;
             }
+        }
+        return false;
+    }
 
-            @Override
-            public void segmentCreated(WireCommands.SegmentCreated segmentCreated) {
-                log.info(requestId, "CreateSegment {} SegmentCreated", qualifiedStreamSegmentName);
-                result.complete(true);
-            }
+    @SneakyThrows(ConnectionFailedException.class)
+    private Boolean transformReplyToBoolean(Reply reply, Class[] classArray) {
+        if (isReplyWanted(reply, classArray)) {
+            return true;
+        }
 
-            @Override
-            public void processingFailure(Exception error) {
-                log.error(requestId, "CreateSegment {} threw exception", qualifiedStreamSegmentName, error);
-                handleError(error, result, type);
-            }
+        closeConnection(reply);
+        if (reply instanceof WireCommands.NoSuchSegment) {
+            throw new NoSuchSegmentException(reply.toString());
+        } else if (reply instanceof WireCommands.WrongHost) {
+            throw new ConnectionFailedException(reply.toString());
+        } else {
+            throw new ConnectionFailedException("Unexpected reply of " + reply + " when expecting a "
+                    + classArray[0].getName());
+        }
+    }
 
-            @Override
-            public void authTokenCheckFailed(WireCommands.AuthTokenCheckFailed authTokenCheckFailed) {
-                result.completeExceptionally(
-                        new WireCommandFailedException(new AuthenticationException(authTokenCheckFailed.toString()),
-                                type, WireCommandFailedException.Reason.AuthFailed));
-            }
-        };
+    public CompletableFuture<Boolean> createSegment(final String scope,
+                                                    final String stream,
+                                                    final long segmentId,
+                                                    final ScalingPolicy policy,
+                                                    String controllerToken,
+                                                    final long clientRequestId) {
+        final String qualifiedStreamSegmentName = getQualifiedStreamSegmentName(scope, stream, segmentId);
+        final Controller.NodeUri uri = getSegmentUri(scope, stream, segmentId);
+        //final long requestId = (clientRequestId == RequestTag.NON_EXISTENT_ID) ? idGenerator.get() : clientRequestId;
+        final long requestId = (clientRequestId == RequestTag.NON_EXISTENT_ID) ? idGenerator.get() : idGenerator.get() + clientRequestId;
```

Comment:
The usage of idGenerator is incorrect as there can be an id collision, the id is used by the connection pooling logic to ensure reposes to every `Flow` are correctly handled. 

The id (type `long`) contains two parts [Flow Id `int` + Sequence Number `int` ] .  
The flow id is used by the connection pooling logic to ensure correct `ReplyProcessor` is invoked. (In this case the ReplyProcessor of the RawClient should be invoked for the given given `Flow Id`.)

The pattern used inside `SegmentMetadataClientImpl`, which is very similar to the pattern this PR implements with the SegmentHelper, can be reused here. This will ensure all the operation performed in the SegmentHelper can be handled by reusing the underlying RawClient(s).

Reply:
this has been fixed in recent commit